### ISSUE TITLE
OCO Function 

### ISSUE THREAD

**Issue by crlsnr** [tony-ho/binance-oco#1]
Thanks for writing this script.  I was in the middle of writing my own OCO bot in python for binance when I found this.

I've done a few cheap test runs of the OCO function and the script seems to be hung up after the buy order and sell-limit orders are FILLED.  The trading pair price clearly went above my targets, but the bot did nothing.  No cancellation of the previous stop-limit order and no new sell-limit order placed at the target.  There were no "trade update.." console messages printed.  The last message printed was the sell-limit order id.  Same results whether I do a buy limit order or a buy market order (price of 0).

test run example:
`binance-oco -p BTCUSDT -a .002 -b 0 -t 8425 -s 8400`

Version 1.2.1
Windows 10 w/ all necessarily dependencies

I'm not as familiar with javascript, but could there be a logical error in here?:

`const placeSellOrder = function () {
    if (stopPrice) {
      placeStopOrder();
    } else if (targetPrice) {
      placeTargetOrder();
    } else {
      process.exit();
    }
  };`

If the user supplies both the stopPrice and the targetPrice, the 2nd else if will never be executed.

Thanks in advance - looking forward to using this.

**Comment by crlsnr** [comment]
@tony-ho Any thoughts on this issue?  Thanks in advance.

**Comment by tony-ho** [comment]
@crlsnr Thanks for raising this issue!

I've identified the underlying cause. It is due to the `newOrderRespType: 'FULL'` flag not being set when placing the sell orders.

See:
https://github.com/jaggedsoft/node-binance-api/blob/master/examples/advanced.md#neworderresptype-example-when-placing-orders

This causes the API response to not include the `type` field, meaning neither of the below paths in `sellOrderComplete` are being executed.

```
    if (response.type === 'STOP_LOSS_LIMIT') {
      stopOrderId = response.orderId;
    } else if (response.type === 'LIMIT') {
      targetOrderId = response.orderId;
    }
```

Since neither `stopOrderId` or `targetOrderId` are set, the trade update logic is bypassed.

This will be fixed in the next release.

**Comment by crlsnr** [comment]
@tony-ho Thanks for taking a look at this.  I did a couple of tests and it's not hanging anymore, but there's some logic that isn't quite right.  Using the same launch parameters as above (different prices obviously) it will successfully buy, set the stop-limit-sell, and cancel the stop-limit-sell if the target price is reached, however it does not create a new sell-limit order at the designated target price.  It simply exits after it cancels the stop-limit-sell.

**Comment by tony-ho** [comment]
@crlsnr Can you send me the command line you used to reproduce the error and the console output?

I tested this with the latest version and it successfully creates the target limit order.

```
$ node binance-oco.js -p XVGBTC -a 500 -b 0 -s 0.0000024 -t 0.00000248
Buy response { symbol: 'XVGBTC',
  orderId: 45739775,
  clientOrderId: 'KweArJBaF5UgrZ1FQ9VW1n',
  transactTime: 1533732197945,
  price: '0.00000000',
  origQty: '500.00000000',
  executedQty: '500.00000000',
  cummulativeQuoteQty: '0.00124000',
  status: 'FILLED',
  timeInForce: 'GTC',
  type: 'MARKET',
  side: 'BUY',
  fills:
   [ { price: '0.00000248',
       qty: '500.00000000',
       commission: '0.00048438',
       commissionAsset: 'BNB',
       tradeId: 17510997 } ] }
order id: 45739775
Sell response { symbol: 'XVGBTC',
  orderId: 45739776,
  clientOrderId: 'Tq4rZiNcoG9qPdkx1BOd9W',
  transactTime: 1533732198142,
  price: '0.00000240',
  origQty: '500.00000000',
  executedQty: '0.00000000',
  cummulativeQuoteQty: '0.00000000',
  status: 'NEW',
  timeInForce: 'GTC',
  type: 'STOP_LOSS_LIMIT',
  side: 'SELL',
  stopPrice: '0.00000240',
  fills: [] }
order id: 45739776
XVGBTC trade update. price: 0.00000247 stop: 0.0000024 target: 0.00000248
XVGBTC trade update. price: 0.00000248 stop: 0.0000024 target: 0.00000248
XVGBTC cancel response: { symbol: 'XVGBTC',
  origClientOrderId: 'Tq4rZiNcoG9qPdkx1BOd9W',
  orderId: 45739776,
  clientOrderId: 'PbMHQE3AAYxPxrMyceLM6X' }
Sell response { symbol: 'XVGBTC',
  orderId: 45739891,
  clientOrderId: 'ycFl09jCL3Vrk3Er9s0EUc',
  transactTime: 1533732472397,
  price: '0.00000248',
  origQty: '500.00000000',
  executedQty: '0.00000000',
  cummulativeQuoteQty: '0.00000000',
  status: 'NEW',
  timeInForce: 'GTC',
  type: 'LIMIT',
  side: 'SELL',
  fills: [] }
order id: 45739891
XVGBTC SELL LIMIT ORDER #45739891 (NEW)
..price: 0.00000248, quantity: 500.00000000
XVGBTC trade update. price: 0.00000247 stop: 0.0000024 target: 0.00000248
```

**Comment by crlsnr** [comment]
Yup no problem.  During this run it did complete the "OCO" orders, but threw an error at the end.  I'll do some more testing to try and reproduce the issue where the new target order was not being placed.

```
$node binance-oco.js -p BTCUSDT -a .002 -b 0 -s 6465 -t 6475
Buy response { symbol: 'BTCUSDT',
  orderId: 144841669,
  clientOrderId: 'xXs1EfRgB5YC2j4XjIVGtZ',
  transactTime: 1533741236723,
  price: '0.00000000',
  origQty: '0.00200000',
  executedQty: '0.00200000',
  cummulativeQuoteQty: '12.94768000',
  status: 'FILLED',
  timeInForce: 'GTC',
  type: 'MARKET',
  side: 'BUY',
  fills:
   [ { price: '6473.84000000',
       qty: '0.00200000',
       commission: '0.00000200',
       commissionAsset: 'BTC',
       tradeId: 61863757 } ] }
order id: 144841669
Sell response { symbol: 'BTCUSDT',
  orderId: 144841673,
  clientOrderId: 'MZhpRMQnW2cAYwPbSvZuCd',
  transactTime: 1533741236971,
  price: '6465.00000000',
  origQty: '0.00199800',
  executedQty: '0.00000000',
  cummulativeQuoteQty: '0.00000000',
  status: 'NEW',
  timeInForce: 'GTC',
  type: 'STOP_LOSS_LIMIT',
  side: 'SELL',
  stopPrice: '6465.00000000',
  fills: [] }
order id: 144841673
BTCUSDT trade update. price: 6474.35000000 stop: 6465 target: 6475
BTCUSDT trade update. price: 6473.28000000 stop: 6465 target: 6475
BTCUSDT trade update. price: 6473.28000000 stop: 6465 target: 6475
BTCUSDT trade update. price: 6473.26000000 stop: 6465 target: 6475
BTCUSDT trade update. price: 6473.20000000 stop: 6465 target: 6475
BTCUSDT trade update. price: 6471.96000000 stop: 6465 target: 6475
BTCUSDT trade update. price: 6471.87000000 stop: 6465 target: 6475
BTCUSDT trade update. price: 6471.70000000 stop: 6465 target: 6475
BTCUSDT trade update. price: 6471.70000000 stop: 6465 target: 6475
BTCUSDT trade update. price: 6471.68000000 stop: 6465 target: 6475
BTCUSDT trade update. price: 6471.64000000 stop: 6465 target: 6475
BTCUSDT trade update. price: 6472.33000000 stop: 6465 target: 6475
BTCUSDT trade update. price: 6473.30000000 stop: 6465 target: 6475
BTCUSDT trade update. price: 6473.30000000 stop: 6465 target: 6475
BTCUSDT trade update. price: 6474.34000000 stop: 6465 target: 6475
BTCUSDT trade update. price: 6474.34000000 stop: 6465 target: 6475
BTCUSDT trade update. price: 6474.35000000 stop: 6465 target: 6475
BTCUSDT trade update. price: 6474.35000000 stop: 6465 target: 6475
BTCUSDT trade update. price: 6474.35000000 stop: 6465 target: 6475
BTCUSDT trade update. price: 6474.79000000 stop: 6465 target: 6475
BTCUSDT trade update. price: 6475.24000000 stop: 6465 target: 6475
BTCUSDT trade update. price: 6475.24000000 stop: 6465 target: 6475
BTCUSDT cancel response: { symbol: 'BTCUSDT',
  origClientOrderId: 'MZhpRMQnW2cAYwPbSvZuCd',
  orderId: 144841673,
  clientOrderId: 'PXQ6JKLRrz05aqwwZ06bMS' }
Sell response { symbol: 'BTCUSDT',
  orderId: 144841725,
  clientOrderId: 'q05GCGhOluvyKRdXDAGh1Q',
  transactTime: 1533741246630,
  price: '6475.00000000',
  origQty: '0.00199800',
  executedQty: '0.00000000',
  cummulativeQuoteQty: '0.00000000',
  status: 'NEW',
  timeInForce: 'GTC',
  type: 'LIMIT',
  side: 'SELL',
  fills: [] }
order id: 144841725
BTCUSDT SELL LIMIT ORDER #144841725 (NEW)
..price: 6475.00000000, quantity: 0.00199800
BTCUSDT cancel error: {"code":-2011,"msg":"UNKNOWN_ORDER"}>
```

**Comment by tony-ho** [comment]
@crlsnr Thanks for sending through the console output.

I've identified the issue. In your output, two trade updates where the price has reached the target occur before the script has cancelled the stop order. The second update causes the script to attempt to cancel the stop order again. This fails with an error and exits the script.

The script has been updated to prevent multiple calls to cancel the same order, and to continue instead of exiting if there is an error.
